src/client/metric/timer.rs

```
@@ -0,0 +1,124 @@
+use super::*;
+use time;
+use time::Tm;
+
+/// A timer metric for tracking elapsed time
+///
+/// Internally uses a `Metric<i64>` with `Semantics::Instant` and `1` time dimension
+pub struct Timer {
+    metric: Metric<i64>,
+    time_scale: Time,
+    start_time: Option<Tm>
+}
+
+impl Timer {
+    /// Creates a new timer metric with given time scale
+    pub fn new(name: &str, time_scale: Time,
+        shorthelp_text: &str, longhelp_text: &str) -> Result<Self, String> {
+
+        let metric = Metric::new(
+            name,
+            0,
+            Semantics::Instant,
+            Unit::new().time(time_scale, 1)?,
+            shorthelp_text,
+            longhelp_text
+        )?;
+
+        Ok(Timer {
+            metric: metric,
+            time_scale: time_scale,
+            start_time: None
+        })
+    }
+
+    /// Starts the timer. Returns an error if the timer is
+    /// already started.
+    pub fn start(&mut self) -> Result<(), String> {
+        if self.start_time.is_some() {
+            return Err("Timer already started!".to_owned());
+        }
+        self.start_time = Some(time::now());
+        Ok(())
+    }
+
+    /// Stops the timer, updates the internal metric, and
+    /// returns the time elapsed since the last `start`. If
+    /// the timer was stopped too early or too late such that
+    /// the internal nanosecond, microsecond or millisecond value
+    /// over/under-flows, then elapsed time isn't updated.
+    /// 
+    /// Returns `0` if the timer wasn't started before.
+    pub fn stop(&mut self) -> io::Result<i64> {
```

Why not return an Err if the timer was not started before